# Predicting NYC Food Deserts EDA

In [35]:
# import necessary packages
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
sns.set_style('darkgrid')
import censusdata
from sodapy import Socrata
import warnings
warnings.filterwarnings('ignore')

# set directory
PROJECT_DIR = "."
IMAGES_PATH = os.path.join(PROJECT_DIR, 'images')
os.makedirs(IMAGES_PATH, exist_ok=True)

# function for saving figures
def save_fig(fig_id, tight_layout=True, fig_extension="png", resolution=300):
    path = os.path.join(IMAGES_PATH, fig_id + "." + fig_extension)
    print("Saving figure", fig_id)
    if tight_layout:
        plt.tight_layout()
    plt.savefig(path, format=fig_extension, dpi=resolution)

## Import data

### NYC Census Tracts Shapefile

In [36]:
# set client for request Socrata API
client = Socrata("data.cityofnewyork.us", app_token='2zRBf7lHUfSBGGX44PZy1XhmT')

# get shapefile from Socrata API
ct_shp = client.get("7igh-afai", limit=20000)

# convert to pandas GeoDataFrame
ct_df = pd.DataFrame.from_records(ct_shp)

### USDA data import

In [37]:
# import usa dataset
usda_df = pd.read_excel('data/DataDownload2015.xlsx', sheet_name='Food Access Research Atlas', header=0)

# create data dict df
usda_data_dict = pd.read_excel('data/DataDownload2015.xlsx', sheet_name='Variable Lookup') 

In [38]:
usda_df.head()

CensusTract    State   County  Urban  POP2010  OHU2010  GroupQuartersFlag  \
0  1001020100   Alabama  Autauga  1      1912     693      0                   
1  1001020200   Alabama  Autauga  1      2170     743      0                   
2  1001020300   Alabama  Autauga  1      3373     1256     0                   
3  1001020400   Alabama  Autauga  1      4386     1722     0                   
4  1001020500   Alabama  Autauga  1      10766    4082     0                   

   NUMGQTRS  PCTGQTRS  LILATracts_1And10  LILATracts_halfAnd10  \
0  0         0.000000  0                  0                      
1  181       0.083410  0                  0                      
2  0         0.000000  0                  0                      
3  0         0.000000  0                  0                      
4  181       0.016812  0                  0                      

   LILATracts_1And20  LILATracts_Vehicle  HUNVFlag  LowIncomeTracts  \
0  0                  0                   0         0                 
1  0                  0                   0         0                 
2  0                  0                   0         0                 
3  0                  0                   0         0                 
4  0                  0                   1         0                 

   PovertyRate  MedianFamilyIncome  LA1and10  LAhalfand10  LA1and20  \
0  10.0         74750               1         1            1          
1  18.2         51875               0         1            0          
2  19.1         52905               1         1            1          
3  3.3          68079               1         1            1          
4  8.5          77819               1         1            1          

   LATracts_half  LATracts1  LATracts10  LATracts20  LATractsVehicle_20  \
0  1              1          0           0           0                    
1  1              0          0           0           0                    
2  1              1          0           0           0                    
3  1              1          0           0           0                    
4  1              1          0           0           1                    

     LAPOP1_10   LAPOP05_10    LAPOP1_20  LALOWI1_10  LALOWI05_10  LALOWI1_20  \
0  1357.480940  1732.225468  1357.480940  322.084612  411.690754   322.084612   
1  483.429683   1410.374828  483.429683   145.141418  475.306031   145.141418   
2  1417.874893  2764.604126  1417.874893  696.636797  1350.176549  696.636797   
3  1363.466885  3651.061015  1363.466885  409.587836  1067.844499  409.587836   
4  2643.095161  7778.396188  2643.095161  623.076555  1913.453614  623.076555   

     lapophalf  lapophalfshare   lalowihalf  lalowihalfshare   lakidshalf  \
0  1732.225468  0.905976        411.690754   0.215319         466.426430    
1  1410.374828  0.649942        475.306031   0.219035         448.163512    
2  2764.604126  0.819628        1350.176549  0.400290         744.891575    
3  3651.061015  0.832435        1067.844499  0.243467         846.613783    
4  7778.396188  0.722496        1913.453614  0.177731         2309.475651   

   lakidshalfshare  laseniorshalf  laseniorshalfshare  lawhitehalf  \
0  0.243947         198.828220     0.103990            1482.881417   
1  0.206527         139.305390     0.064196            412.291349    
2  0.220839         346.203097     0.102640            2114.851375   
3  0.193026         766.521491     0.174766            3395.061920   
4  0.214516         839.885471     0.078013            6298.710335   

   lawhitehalfshare  lablackhalf  lablackhalfshare  laasianhalf  \
0  0.775566          184.319106   0.096401          12.718121     
1  0.189996          945.317796   0.435630          4.000000      
2  0.626994          528.050101   0.156552          9.031083      
3  0.774068          170.065452   0.038775          14.874754     
4  0.585056          1000.678496  0.092948          208.982771    

   laasianhalfshare  lanhopihalf  lanhopihalfs

In [39]:
pd.options.display.max_rows = None
pd.options.display.max_columns = None
pd.set_option('display.max_colwidth', -1)
usda_data_dict

Field  \
0    CensusTract            
1    State                  
2    County                 
3    Urban                  
4    POP2010                
5    OHU2010                
6    GroupQuartersFlag      
7    NUMGQTRS               
8    PCTGQTRS               
9    LILATracts_1And10      
10   LILATracts_halfAnd10   
11   LILATracts_1And20      
12   LILATracts_Vehicle     
13   HUNVFlag               
14   LowIncomeTracts        
15   PovertyRate            
16   MedianFamilyIncome     
17   LA1and10               
18   LAhalfand10            
19   LA1and20               
20   LATracts_half          
21   LATracts1              
22   LATracts10             
23   LATracts20             
24   LATractsVehicle_20     
25   LAPOP1_10              
26   LAPOP05_10             
27   LAPOP1_20              
28   LALOWI1_10             
29   LALOWI05_10            
30   LALOWI1_20             
31   lapophalf              
32   lapophalfshare         
33   lalowihalf             
34   lalowihalfshare        
35   lakidshalf             
36   lakidshalfshare        
37   laseniorshalf          
38   laseniorshalfshare     
39   lawhitehalf            
40   lawhitehalfshare       
41   lablackhalf            
42   lablackhalfshare       
43   laasianhalf            
44   laasianhalfshare       
45   lanhopihalf            
46   lanhopihalfshare       
47   laaianhalf             
48   laaianhalfshare        
49   laomultirhalf          
50   laomultirhalfshare     
51   lahisphalf             
52   lahisphalfshare        
53   lahunvhalf             
54   lahunvhalfshare        
55   lasnaphalf             
56   lasnaphalfshare        
57   lapop1                 
58   lapop1share            
59   lalowi1                
60   lalowi1share           
61   lakids1                
62   lakids1share           
63   laseniors1             
64   laseniors1share        
65   lawhite1               
66   lawhite1share          
67   lablack1               
68   lablack1share          
69   laasian1               
70   laasian1share          
71   lanhopi1               
72   lanhopi1share          
73   laaian1                
74   laaian1share           
75   laomultir1             
76   laomultir1share        
77   lahisp1                
78   lahisp1share           
79   lahunv1                
80   lahunv1share           
81   lasnap1                
82   lasnap1share           
83   lapop10                
84   lapop10share           
85   lalowi10               
86   lalowi10share          
87   lakids10               
88   lakids10share          
89   laseniors10            
90   laseniors10share       
91   lawhite10              
92   lawhite10share         
93   lablack10              
94   lablack10share         
95   laasian10              
96   laasian10share         
97   lanhopi10              
98   lanhopi10share         
99   laaian10               
100  laaian10share          
101  laomultir10            
102  laomultir10share       
103  lahisp10               
104  lahisp10share          
105  lahunv10               
106  lahunv10share          
107  lasnap10               
108  lasnap10share          
109  lapop20                
110  lapop20share           
111  lalowi20               
112  lalowi20share          
113  lakids20               
114  lakids20share          
115  laseniors20            
116  laseniors20share       
117  lawhite20              
118  lawhite20share         
119  lablack20              
120  lablack20share         
121  laasian20              
122  laasian20share         
123  lanhopi20              
124  lanhopi20share         
125  laaian20               
126  laaian20share          
127  laomultir20            
128  laomultir20share       
129  lahisp20               
130  lahisp20share          
131  lahunv20               
132  lahunv20share          
133  lasnap20               
134  lasnap20share          
135  TractLOWI              
136  TractKids              
137  TractSeniors 

In [40]:
usda_data_dict.to_csv('data/usda_data_dict.csv')

### Data cleaning

Filter out data by state for New York only.

In [41]:
usda_df = usda_df.loc[usda_df['State'] == 'New York']


Filter out data for 5 boroughs

In [42]:
# create filter for boolean mask to isolate 5 boroughs
five_boroughs =  (usda_df['County'] == 'Bronx') | (usda_df['County'] == 'Kings') | (usda_df['County'] == 'New York') |\
        (usda_df['County'] == 'Queens') | (usda_df['County'] == 'Richmond')

# create usda nyc df
usda_df_nyc = usda_df.loc[five_boroughs]
usda_df_nyc.to_csv('data/usda_df_nyc.csv')

In [43]:
# look at resultant shape of df
usda_df_nyc.shape


(2165, 147)

There are 2165 rows (census tracts) in New York City 5 boroughs and 147 columns.

In [44]:
# get value counts for columsn
usda_df_nyc['County'].value_counts()

Kings       760
Queens      669
Bronx       339
New York    288
Richmond    109
Name: County, dtype: int64

Look at value counts for target column

In [45]:
# isolate target column to see classification
usda_df_nyc['LILATracts_halfAnd10'].value_counts()

0    2134
1    31  
Name: LILATracts_halfAnd10, dtype: int64

In [46]:
nyc_food_desert = usda_df['LILATracts_halfAnd10'] == 1
nyc_food_desert_df = usda_df.loc[nyc_food_desert]
nyc_food_desert_df.shape


(545, 147)

In [47]:
# look at dtypes of each column
usda_df_nyc.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2165 entries, 43145 to 47000
Columns: 147 entries, CensusTract to TractSNAP
dtypes: float64(112), int64(33), object(2)
memory usage: 2.4+ MB


147 columns 112 floats, 33 ints and 2 objects

Drop unnecesary columns, only concerned with data of 1/2 mile access.

In [52]:
usda_df_nyc.drop(usda_df_nyc.iloc[:,57:135], inplace=True, axis=1)

In [56]:
usda_df_nyc.columns

Index(['CensusTract', 'State', 'County', 'Urban', 'POP2010', 'OHU2010',
       'GroupQuartersFlag', 'NUMGQTRS', 'PCTGQTRS', 'LILATracts_1And10',
       'LILATracts_halfAnd10', 'LILATracts_1And20', 'LILATracts_Vehicle',
       'HUNVFlag', 'LowIncomeTracts', 'PovertyRate', 'MedianFamilyIncome',
       'LA1and10', 'LAhalfand10', 'LA1and20', 'LATracts_half', 'LATracts1',
       'LATracts10', 'LATracts20', 'LATractsVehicle_20', 'LAPOP1_10',
       'LAPOP05_10', 'LAPOP1_20', 'LALOWI1_10', 'LALOWI05_10', 'LALOWI1_20',
       'lapophalf', 'lapophalfshare', 'lalowihalf', 'lalowihalfshare',
       'lakidshalf', 'lakidshalfshare', 'laseniorshalf', 'laseniorshalfshare',
       'lawhitehalf', 'lawhitehalfshare', 'lablackhalf', 'lablackhalfshare',
       'laasianhalf', 'laasianhalfshare', 'lanhopihalf', 'lanhopihalfshare',
       'laaianhalf', 'laaianhalfshare', 'laomultirhalf', 'laomultirhalfshare',
       'lahisphalf', 'lahisphalfshare', 'lahunvhalf', 'lahunvhalfshare',
       'lasnaphalf', 'lasna